trusted_application_demo/client_application/aes_certify_messaging/publisher/aes_publisher.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+* Copyright (c) 2018, ARM Limited.
+*
+* SPDX-License-Identifier: Apache-2.0
+*/
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "aes_custom_interface/srv/get_symm_secret.hpp"
+#include "aes_custom_interface/msg/aes_message_certify.hpp"
+
+#include <string>
+
+extern "C" {
+  #include <teec_utils.h>
+  #include <rsa_api.h>
+  #include <aes_api.h>
+  #include <hmac_api.h>
+  #include <tee_client_api.h>
+  #include <ta_public.h>
+}
+
+// Max number of byte for a message.
+#define PUBLISHER_MAX_MESSAGE_SIZE      18
+// Number of messages that will be sent by the publisher.
+#define PUBLISHER_MAX_NUMBER_MESSAGE   100
+
+using namespace std;
+using namespace placeholders;
+using namespace aes_custom_interface::srv;
+using namespace aes_custom_interface::msg;
+
+// Minimal publisher inherite from class Node
+class AesPublisher : public rclcpp::Node
+{
+public:
+  AesPublisher(string topic_name, string service_name)
+  // The following argument are declared for the superclass
+  // Create a new node with name minimal_publisher, init variable count_
+  : Node("publisher_and_server") {
+    publisher = this->create_publisher<AesMessageCertify>(topic_name,
+                                                   rmw_qos_profile_default);
+    TEEC_UUID uuid = TA_TRUSTED_UUID;
+
+    teec_exitOnFailure(&ctx, &sess,
+                       teec_openSession(uuid, &ctx, &sess)
+                      );
+
+    allocate_memory_ta();
+
+    // Create the AES secret and encrypt it with RSA
+    teec_exitOnFailure(&ctx, &sess,
+                       aes_generateAndEncryptSecret(&sess,
+                                                    &aes_key_rsa_encrypted_shm,
+                                                    RSA_API_SIZE_ENCRYPTED)
+                      );
+
+    teec_exitOnFailure(&ctx, &sess,
+                       rsa_certify(&sess,
+                                   &aes_key_rsa_encrypted_shm,
+                                   aes_key_rsa_encrypted_shm.size,
+                                   &sha1_rsa_certified_shm)
+                      );
+
+    // Create the service used to ask for the secret
+    auto bind_service_handler = bind(&AesPublisher::handle_service,
+                                     this, _1, _2, _3);
+    publisher_service = this->create_service<GetSymmSecret>(
+                                  service_name,
+                                  bind_service_handler,
+                                  rmw_qos_profile_default);
+
+    // Message ID used for all the messages.
+    message_id = 0;
+
+    RCLCPP_INFO(this->get_logger(),
+            "Publisher is fully initialized with \n"
+            "<topic>: \"%s\" and <service>: \"%s\"",
+            topic_name.c_str(), service_name.c_str())
+  }
+
+  bool send_msg(void) {
+    // Create the message at the correct memory location
+    char* clear_message_cast = (char*) clear_message_shm.buffer;
+    snprintf(clear_message_cast, PUBLISHER_MAX_MESSAGE_SIZE,
```

Comment:
snprintf is considered unsafe. you can use [`rcutils_snprintf`](https://github.com/ros2/rcutils/blob/3a8957008a5e05f30fe4060d8cb605cb6aceab82/include/rcutils/snprintf.h) instead